File: cpp/2929.cpp
#include <iostream>
#include <string>
#include <cmath>
#include <stack>

using namespace std;

class Stack
{
    stack<pair<int, int>> my_stack;

    public:
        int count = 0;

        void push(int value)
        {
            int curr_min;

            if (count == 0) {
                curr_min = value;
            } else {
                curr_min = min(my_stack.top().first, value);
            }

            count++;
            my_stack.push(make_pair(curr_min, value));
        }

        void pop()
        {
            count--;
            my_stack.pop();
        }

        int get_min()
        {
            return my_stack.top().first;
        }
};

int main()
{
    cin.tie(0)->sync_with_stdio(false);

    Stack stack;
    string op;
    int n, v;
    cin >> n;

    for (n; n > 0; n--) {
        cin >> op;

        if (op == "PUSH") {
            cin >> v;
            stack.push(v);
        } else if (stack.count == 0) {
            cout << "EMPTY\n";
        } else if (op == "POP") {
            stack.pop();
        } else {
            cout << stack.get_min() << '\n';
        }
    }

    return 0;
}

File: py/2929.py
from collections import deque
from sys import stdin, stdout

class Stack(object):
    def __init__(self):
        self.stack = deque()
        self.count = 0


    def push(self, value):
        if self.count == 0:
            curr_min = value
        else:
            curr_min = min(value, self.stack[-1][0])

        self.count += 1
        self.stack.append((curr_min, value))


    def pop(self):
        self.count -= 1
        self.stack.pop()


    def get_min(self):
        return self.stack[-1][0]


def main():
    n = int(input())
    stack = Stack()

    for _ in range(n):
        op = stdin.readline().split()

        if op[0] == 'PUSH':
            stack.push(int(op[1]))
        elif stack.count == 0:
            print('EMPTY')
        elif op[0] == 'POP':
            stack.pop()
        else:
            stdout.write(str(stack.get_min()) + '\n')


if __name__ == '__main__':
    main()